sound/hda/intel-nhlt.c

```diff
@@ -43,6 +43,40 @@ int intel_nhlt_get_dmic_geo(struct device *dev, struct nhlt_acpi_table *nhlt)
 		if (epnt->linktype == NHLT_LINK_DMIC) {
 			cfg = (struct nhlt_dmic_array_config  *)
 					(epnt->config.caps);
```

Comment:
I found different usage in another source.   

epnt->config is defined as
````C
struct nhlt_specific_cfg {
        u32 size;
        u8 caps[];
} __packed;
````
we can't covert it to struct nhlt_dmic_array_config directly only if nhlt_specific_cfg->size == 3. And also I found other usage deal with nhlt_specific_cfg->size == 4 for microphone count.
````C
struct nhlt_device_specific_config {
        u8 virtual_slot;
        u8 config_type;
} __packed;

struct nhlt_dmic_array_config {
        struct nhlt_device_specific_config device_config;
        u8 array_type;
} __packed;
````
The correct usage should be
if ( nhlt_specific_cfg->size >= 1) check virtual slot;
if ( nhlt_specific_cfg->size >= 2) check config_type;
if ( nhlt_specific_cfg->size >= 3) check array_type;
if ( nhlt_specific_cfg->size == 4) check microphone count;

Reply:
In the public spec, size if defined to be at least 2, so the first two cases seems like very old backwards-compatibility logic. Cases three and four, seem aligned with @plbossart 's patch here. The spec does define three types of specific-data types, "array", "basic" and "renderfeedback". We could parse these direcly and flag any unknown "config_type"'s.